### ISSUE TITLE
Multi-tenancy input request

### ISSUE THREAD

**Issue by roeltm** [matrix-org/synapse#11872]
Hi, 

We are going to try and make Synapse multi-tenant. To do this we wrote a plan with how we intend to do this. We'd like some input from those with intimate knowledge of synapse to see if we missed something and to check if getting this merged here is possible. Would be a bit of a waste of spending a lot of time on this and then not being able to get it merged.

Can any of you guys give this a read and comment: https://gist.github.com/roeltm/1556201690ea5c10cc965c0b34c1b3cd

Thanks in advance.

**Comment by richvdh** [comment]
related: #8885

**Comment by richvdh** [comment]
Hi @roeltm, thanks for your work in preparing this.

My initial thought reading this: it looks like you're proposing to have completely separate postgres/sqlite databases (with their own connection pools?) for each domain, and separate copies of the in-memory caches for each domain. In short, for a public room where you have users from several domains in the same room, you're going to end up duplicating work and storage for each domain.

So my initial thought is: this seems to be an awful lot of work to achieve something that looks broadly similar to simply running several Synapse processes. Or am I missing something?

**Comment by roeltm** [comment]
> So my initial thought is: this seems to be an awful lot of work to achieve something that looks broadly similar to simply running several Synapse processes. Or am I missing something?

True to a certain degree. Let me explain our situation, that might clear something up.

The company I work for is trying to provide advanced chat functionalities for our customers (in healthcare). Our initial starting point is simply running seperate instances/tenants for every customer since most of the communication will take place between the employees of that customer. After that we are going to look at combining them in a private federation in order to support communication between customers.

Since we're talking about a fairly large number of customer (>1000), having to manage multiple instances, having to add new customers, keeping everything up to date etc. is going to be quite the task. That is why we prefer to be able to run synapse in multi-tenant, multi-database mode. We're also talking about quite a few of these customers having only <100 employees. So having a single instance for that customer is quite the overkill.

**Comment by richvdh** [comment]
ok, thanks for the explanation. My feeling is that we'd be unlikely to merge such a change to mainline, since it would add quite a lot of complexity without being very useful in most cases. I'll flag it for discussion with the team though.

**Comment by roeltm** [comment]
> since it would add quite a lot of complexity without being very useful in most cases

I understand that the complexity will definitely be impacted. But I don't think I agree with the second part. If you want to talk about adoption of the standard, having software companies including chat capabilities to their customers using Matrix is going to be a big part. Since those customers are usually seperated from the federation (to a certain degree) and would probably like their own domain, the software companies will be required to run a matrix server for a large amount of customers. Supporting multi-tenancy would greatly reduce the effort/resources for these parties.

**Comment by roeltm** [comment]
Also: element offers matrix hosting. Won't that also benefit from being able to run multi-tenant to reduce resource and maintanance time?

**Comment by ShadowJonathan** [comment]
Related: https://github.com/matrix-org/matrix-doc/pull/3383

**Comment by richvdh** [comment]
Hi @roeltm, thanks again for all your work here. I've discussed this with the rest of the team and I think I speak for all of them in what I'm about to say.

> Also: element offers matrix hosting. Won't that also benefit from being able to run multi-tenant to reduce resource and maintanance time?

Certainly, EMS would very much benefit from a reduction in resource usage. However, my concern is that your approach will do little to materially reduce resource usage, versus simply running multiple copies of Synapse. All of the I/O, database, and memory requirements remain. (It's true that you would get some benefit from not duplicating the python codebase between different Synapse processes, but that is, relatively speaking, a tiny part of Synapse's memory footprint).

In terms of maintenance time: deployments like yours or EMS are going to end up having to manage farms of servers and instances, requiring an orchestration tool such as kubernetes or ansible, in any case. So the incremental cost of having to restart a couple more processes is pretty small.

My main fear here is that, once we merge a change like this into the main synapse codebase, it becomes very hard to unpick it again. Assuming at some point in the future we implement a vhosting solution which allows a synapse process to host multiple domains *without* duplicating the database and in-memory caches, we then have to support *both* solutions, which becomes very complicated.

In short, my answer remains the same: I don't think we'd be happy merging this into Synapse. I'm sorry that's not the answer you want.

**Comment by roeltm** [comment]
Thanks for taking the time to discuss it and explaining your reasoning 👍

**Comment by ShadowJonathan** [comment]
@roeltm FTR, i personally think your idea (multi-tenancy and vhosting) is sound in general, and may be interesting enough to build a whole new homeserver around.

I don't know if you'd want that, or have the time/resources to do that, but if you do, or find it interesting, i have some ideas about it you may find interesting, if you want to discuss them :)

In general though, i recommend hanging around on [`#homeservers-dev`](https://matrix.to/#/#homeservers-dev:matrix.org) on matrix for those kinds of discussions.

**Comment by ShadowJonathan** [comment]
@dulguun0225 not at the moment that I know of, not publicly and/or announced in TWIM (this week in matrix)

**Comment by dulguun0225** [comment]
@ShadowJonathan I'm thinking of making a half baked matrix homeserver that can make use of already made bridges. I'm not sure how long that takes. Is it too big of a task for a one human? How many man hours would it take?

**Comment by ShadowJonathan** [comment]
I can't say that, I recommend joining [`#homeservers-dev`](https://matrix.to/#/#homeservers-dev:matrix.org) and discussing this there.
